design/open/delta-diff.md

```diff
@@ -0,0 +1,168 @@
+This is almost where we want to be, except that we'll need to somehow validate that the lakeFS server and the executable are using the same "API version" to communicate, so that the output of the binary would match the expected one from lakeFS. In addition, I would like to decrease the amount of deserialization needed to be implemented to interpret the returned DTO (maintainability and extensibility-wise).
+
+### Chosen architecture
+
+#### The Microkernel/Plugin Architecture
+
+The Microkernel/Plugin architecture is composed of two entities: a single "core system" and multiple "plugins".  
+In our case, the lakeFS server act as the core system, and the different diff implementations, including the Delta diff implementation, will act as plugins.  
+We'll use `gRPC` as the transport protocol, which makes the language of choice almost immaterial (due to the use of protobufs as the transferred data format)
+as long as it's self-contained (theoretically it can also be not system-runtime-dependent but then the cost will be an added requirement for running lakeFS- runtime support).
+
+#### Plugin system consideration
+
+* The plugin system should be flexible enough such that it won't impose a language restriction, so that in the future we could have lakeFS users take part in plugins creation (not restricted to diffs).
+* The plugin system should support multiple OSs.
+* The plugin system should be easy enough to write plugins.
+
+#### Hashicorp Plugin system
+
+Hashicorp's battle-tested [`go-plugin`](https://github.com/hashicorp/go-plugin) system is a plugin system over RPC (used in Terraform, Vault, Consul, and Packer).
+Currently, it's only designed to work over a local network.  
+Plugins can be written and consumed in almost every major language. This is achieved by supporting [gRPC](http://www.grpc.io/) as the communication protocol.
+* The plugin system works by launching subprocesses and communicating over RPC (both `net/rpc` and `gRPC` are supported).
+  A single connection is made between any plugin and the core process. For gRPC-based plugins, the HTTP2 protocol handles [connection multiplexing](https://freecontent.manning.com/animation-http-1-1-vs-http-2-vs-http-2-with-push/).
+* The plugin and core system are separate processes, which means that a crash of a plugin won't cause the core system to crash (lakeFS in our case).
+
+![Microkernel architecture overview](diagrams/microkernel-overview.png)
+[(excalidraw file)](diagrams/microkernel-overview.excalidraw)
+
+---
+
+## Implementation
+
+![Delta Diff flow](diagrams/delta-diff-flow.png)
+[(excalidraw file)](diagrams/delta-diff-flow.excalidraw)
+
+#### DiffService
+
+The DiffService will be an internal component in lakeFS which will serve as the Core system.  
+In order to realize which diff plugins are available, we shall use new configuration values:
+1. `LAKEFS_PLUGINS_LOCATION`
```

Comment:
What if there would be additional plugins?

Reply:
It's better if we use the same prefix for all configuration values related to plugins.

Reply:
True.  So everything below should also be `LAKEFS_PLUGINS_LOCATIONS`.  That means that everything lives inside the same configuration object.

Also as specificed this might not really be a plugin: I don't know whether Viper will let us configure an "any object" and be able to pass it usefully to the plugin during load.

If this turns out to be impossible, I would suggest configuring a separate "plugins file"; that one can be a general YAML/JSON object with very few parsed fields, and then each plugin gets its configuration subobject from that file (and presumably parses it using mapstructure in Go, or libwhatever in other languages).